### DISCUSSION TITLE
Lazy properties having types rendered as `any`

### DISCUSSION THREAD

**Discussion by atoff** [spatie/laravel-data#442] [Q&A] [answered]
Hello!

I've been trying to migrate some of my existing API resources to this package to take advantage of typescript generation via `spatie/typescript-transformer`. [This docs page](https://spatie.be/docs/laravel-data/v3/advanced-usage/typescript) seems to imply that by typing a property to Lazy, it will automatically get rendered by the transformer as a nullable type. However, what actually seems to be happening is it is interpreting the `Lazy` as a typescript `any` type:
**ChartOverview Data Object**
```php

<?php

namespace App\Data;

use App\Models\Chart;
use Spatie\LaravelData\Data;
use Spatie\LaravelData\DataCollection;
use Spatie\LaravelData\Lazy;

/** @typescript */
class ChartOverviewData extends Data
{
    public function __construct(
        public string $id,
        public string $name,
        /** @var Lazy|\App\Data\ChartMetaData[] */
        public Lazy|DataCollection $meta,
    ) {
    }

    public static function fromModel(Chart $chart): self
    {
        return self::from([...$chart->toArray(),
            'meta' => Lazy::whenLoaded('meta', $chart, fn () => ChartMetaData::collection($chart->meta)),
        ]);
    }
}
```
**Generated Type from ModuleWriter on `spatie/typescript-transformer`**
```ts
export type ChartOverviewData = {
id: string;
name: string;
meta: any | Array<ChartMetaData>;
};
```

Has anyone else got experience of this, or is there something I am doing wrong? Thanks!

**Comment by rubenvanassche** [comment]
Could you try replacing:

```
/** @var Lazy|\App\Data\ChartMetaData[] */
```

With:

```
#[DataCollectionOf(ChartMetaData::class)]
```

I think that should fix the issue, I also want to remove support for these annotations in the future since they make things way to complicated, attributes FTW!

**Comment by KKSzymanowski** [comment, marked as answer]
Replacing `DtoTransformer` with `DataTypeScriptTransformer` in the `typescript-transformer` config worked for me.

```diff
--- a/config/typescript-transformer.php
+++ b/config/typescript-transformer.php
@@ -1,5 +1,7 @@
 <?php
 
+use Spatie\LaravelData\Support\TypeScriptTransformer\DataTypeScriptTransformer;
+
 return [
     /*
      * The paths where typescript-transformer will look for PHP classes
@@ -30,7 +32,7 @@
         Spatie\LaravelTypeScriptTransformer\Transformers\SpatieStateTransformer::class,
         Spatie\TypeScriptTransformer\Transformers\EnumTransformer::class,
         Spatie\TypeScriptTransformer\Transformers\SpatieEnumTransformer::class,
-        Spatie\LaravelTypeScriptTransformer\Transformers\DtoTransformer::class,
+        DataTypeScriptTransformer::class,
     ],

```

**Reply by atoff** [reply]
Thank you so much! This absolutely did sort it out

**Reply by KKSzymanowski** [reply]
@rubenvanassche @freekmurze Maybe this should go in the docs? I didn't find it there, found it by trial and error.

**Reply by KKSzymanowski** [reply]
Ok nvm, found it in the docs: https://spatie.be/docs/laravel-data/v3/advanced-usage/typescript#content-installation-of-extra-package

**Reply by phil-hudson** [reply]
hey - if we get rid of the DTO transformer, will we need to migrate any old style annotations to attributes? thanks!
